packages/dd-trace/src/log/writer.js

```diff
@@ -0,0 +1,112 @@
+'use strict'
+
+const { storage } = require('../../../datadog-core')
+const { getChannelLogLevel, debugChannel, infoChannel, warnChannel, errorChannel } = require('./channels')
+
+const defaultLogger = {
+  debug: msg => console.debug(msg), /* eslint-disable-line no-console */
+  info: msg => console.info(msg), /* eslint-disable-line no-console */
+  warn: msg => console.warn(msg), /* eslint-disable-line no-console */
+  error: msg => console.error(msg) /* eslint-disable-line no-console */
+}
+
+let enabled = false
+let logger = defaultLogger
+let logLevel = getChannelLogLevel()
+
+function withNoop (fn) {
+  const store = storage.getStore()
+
+  storage.enterWith({ noop: true })
+  fn()
+  storage.enterWith(store)
+}
+
+function unsubscribeAll () {
+  debugChannel.unsubscribe(onDebug)
+  infoChannel.unsubscribe(onInfo)
+  warnChannel.unsubscribe(onWarn)
+  errorChannel.unsubscribe(onError)
+}
+
+function toggleSubscription (enable) {
+  unsubscribeAll()
+
+  if (enable) {
+    if (debugChannel.logLevel >= logLevel) {
+      debugChannel.subscribe(onDebug)
+    }
+    if (infoChannel.logLevel >= logLevel) {
+      infoChannel.subscribe(onInfo)
+    }
+    if (warnChannel.logLevel >= logLevel) {
+      warnChannel.subscribe(onWarn)
+    }
+    if (errorChannel.logLevel >= logLevel) {
+      errorChannel.subscribe(onError)
+    }
+  }
+}
+
+function toggle (enable, level) {
+  if (level !== undefined) {
+    logLevel = getChannelLogLevel(level)
+  }
+  enabled = enable
+  toggleSubscription(enabled)
+}
+
+function use (newLogger) {
+  if (newLogger && newLogger.debug instanceof Function && newLogger.error instanceof Function) {
+    logger = newLogger
+  }
+}
+
+function reset () {
+  logger = defaultLogger
+  enabled = false
+  logLevel = getChannelLogLevel()
+  toggleSubscription(false)
+}
+
+function onError (err) {
+  if (enabled) {
+    if (err instanceof Function) {
+      err = err()
```

Comment:
Why do this function to result conversion here rather than another `processMsg(...)` at the publishing side?

Reply:
I kept the original logic but I'm going to treat them all the same and `processMsg(...)` at the publishing side for error too.